test: added a flow that deploy and tests a mock policy

ctschacher opened the pull request:
**Description**

Added a test that creates a new API (as admin user), adds keyless plan, adds a Mock policy and verifies the response to the mocked endpoint


<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-qyzsdkljkx.chromatic.com)
<!-- Storybook placeholder end -->
<!-- UI placeholder -->
🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.blob.core.windows.net/test-mock-policy/index.html)
_Notes_: The deployed app is linked to the management API of the Element Zero team's environment.
<!-- UI placeholder end -->

ctschacher committed:
test: added a flow that deploy and tests a mock policy

ytvnr commented on gravitee-apim-cypress/cypress/commands/management/api-management-commands.ts:
```diff
@@ -81,3 +81,13 @@ export function stopApi(auth: BasicAuthentication, apiId: string, failOnStatusCo
   });
 }
+
+export function updateApi(auth: BasicAuthentication, apiId: string, apiUpdate, failOnStatusCode = false) {
```
do not forget the type on apiUpdate

ctschacher replied:
Ok, I'll add that

ytvnr commented on gravitee-apim-cypress/cypress/integration/rest-api/apis/mock-policy-flow.ts:
```diff
@@ -0,0 +1,75 @@
+  before(() => {
+    const fakeApi: Api = ApiFakers.api();
+    const fakePlan = PlanFakers.plan({ security: SecurityType.KEY_LESS });
+    createApi(ADMIN_USER, fakeApi)
```
i would prefer to do it as an API_PUBLISHER

ctschacher replied:
Ok, that's true, good hint!

ytvnr commented on gravitee-apim-cypress/cypress/integration/rest-api/apis/mock-policy-flow.ts:
```diff
@@ -0,0 +1,75 @@
+      .created()
+      .then((response) => {
+        createdApi = response.body;
+        createPlan(ADMIN_USER, createdApi.id, fakePlan)
```
same

ytvnr commented on gravitee-apim-cypress/cypress/integration/rest-api/apis/mock-policy-flow.ts:
```diff
@@ -0,0 +1,75 @@
+      mockPolicyApi.paths = apiFixture.paths;
+      updateApi(ADMIN_USER, createdApi.id, mockPolicyApi).ok();
+      deployApi(ADMIN_USER, createdApi.id).ok();
+      startApi(ADMIN_USER, createdApi.id).noContent();
```
same for the roles -> API_PUBLISHER

ytvnr commented on gravitee-apim-cypress/cypress/integration/rest-api/apis/mock-policy-flow.ts:
```diff
@@ -0,0 +1,75 @@
+  after(() => {
+    deletePlan(ADMIN_USER, createdApi.id, createdPlan.id).noContent();
+    stopApi(ADMIN_USER, createdApi.id).noContent();
+    deleteApi(ADMIN_USER, createdApi.id).noContent();
```
same for roles

ytvnr commented on gravitee-apim-cypress/cypress/integration/rest-api/apis/mock-policy-flow.ts:
```diff
@@ -0,0 +1,75 @@
+      mockPolicyApi.version = createdApi.version;
+      mockPolicyApi.visibility = createdApi.visibility;
+      mockPolicyApi.path_mappings = apiFixture.path_mappings;
+      mockPolicyApi.paths = apiFixture.paths;
```
```suggestion
      let mockPolicyApi = {
        name: createdApi.name,
        description: createdApi.description
        ....
      };
```

ctschacher replied:
I agree, that's much better :)

ytvnr commented on gravitee-apim-cypress/cypress/integration/rest-api/apis/mock-policy-flow.ts:
```diff
@@ -0,0 +1,75 @@
+    });
+  });
+
+  it('should successfully call the mocked API endpoint', { retries: 50 }, function () {
```
`retries` is only available when invoked via the mocha CLI, so I don't know how it's done internally in cypress

more over, in my opinion, we have to find a better way to do that. `retries` should not be at `it()` level but at request's one directly.

ctschacher replied:
Ok, thanks for the hint.
By now I found a different solution for that. The next update of the PR will have a custom Cypress command to deal with request that needs retries.

ytvnr commented on gravitee-apim-cypress/cypress/fixtures/json/updateApi_mockPolicy_example.json:
```diff
@@ -0,0 +1,74 @@
+  "services": {},
+  "properties": [],
+  "tags": [],
+  "picture_url": "http://localhost:8083/management/organizations/DEFAULT/environments/DEFAULT/apis/5f697c6b-635a-440e-a97c-6b635a340ed9/picture?hash=1641283944514",
```
forbidden ! 
not sure you will always run your test with localhost:8083 ;)

ctschacher replied:
I get your point. Ok, I'll just remove all the parts that not needed, including that URL.

marcambier reviewed (commented):
Nothing more to add than @ytvnr 

I think using the `{ retries: 50 }` is actually not a good way to ensure API has been deployed. (I may ask, why 50 ?).
I'll get a look on my side to find a better way of doing this.

ytvnr commented on gravitee-apim-cypress/cypress/fixtures/json/updateApi_mockPolicy_example.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "paths": {
+    "/": [
+      {
+        "methods": ["GET"],
+        "description": "Description of the Mock Gravitee Policy",
+        "enabled": true,
+        "mock": {
+          "headers": [
+            {
+              "name": "test-value",
+              "value": "value123"
+            }
+          ],
+          "status": "200",
+          "content": "{\n\"message\" : \"This is a mocked response\"\n}"
+        }
+      }
+    ]
+  },
+  "path_mappings": ["/"]
+}
```
why using paths ? 
I think it matches the v1 of api definition but we should work with v2 (using plans, floxs, etc)

ctschacher replied:
Ok, right. We should prioritise the API v2 and I'll add a test for that. However, since we'll keep v1 for awhile (until 3.20) and the test is now already written I'll also leave that test there for now.

ytvnr commented on gravitee-apim-cypress/cypress/model/apis.ts:
```diff
@@ -220,3 +224,239 @@ export enum ApiPrimaryOwnerType {
+  metadata?: any;
+  background?: string;
+  proxy?: Proxy;
+  paths?: Rule[];
```
I think it's the v1 style

ytvnr commented on gravitee-apim-cypress/cypress/model/apis.ts:
```diff
@@ -220,3 +224,239 @@ export enum ApiPrimaryOwnerType {
+  background_url?: string;
+}
+
+export interface ResponseTemplate {
```
for all the following types, I dislike the fact that we have everything in one and only file

ctschacher replied:
Yes, I know and I understand the initial motivation to split it into different files but I find it very difficult to assign every interface to a certain file. Some interfaces are used in different areas (files). I could try to assign them to other files but it partly feels quite arbitrary to do so.

ytvnr replied:
does not need one file per interface, but at least we can goup them the best we can
everything relating to logging for an api in api-logging.ts, to proxy in api-proxy.ts
wdyt ?

ctschacher replied:
Alright, as discussed, I get your point and will try my best to put them into right files.

Commit:
test(cypress): improve gateway request from cypress tests

Simplify the way of requesting gateway from cypress tests.

The requestGateway retries calls to gateway with delay, until the "validWhen" condition is met.
By default, this condition is the HTTP 200 OK gateway response, as is most cases, consumer wants to ensure his API has been deployed.
But it can be overriden by caller if needed.

Delays and maximum wait time can also be overriden by caller if needed.
